TrustedTypes: Worker()/SharedWorker() constructors (#42073)
mdn/content: files/en-us/web/api/worker/worker/index.md

### Module and classic workers

Workers can be classic workers or module workers.
The type of worker affects the worker constructor options, how the worker script is fetched, and how it is executed.

The code below shows two ways you can construct a classic worker, and also how you specify the `type` of `"module"` to create a module worker.
In both cases the script must be same-origin with the loading document and is resolved relative to the location of the launching document.

```js
// Construct a classic worker
const worker = new Worker("worker_classic.js");

// Construct a classic worker specifying the tyype
const worker = new Worker("worker_classic.js", {
  type: "classic",
});

// Construct a module worker
const worker = new Worker("worker_module.js", {
  type: "module",
});
```

Module workers and their dependencies are loaded and executed using ECMAScript module semantics:

- Fetched using [CORS](/en-US/docs/Web/HTTP/Guides/CORS)
- All modules are resolved before any code is executed
- Must be served with the media type `Content-Type: text/javascript`
- Executed in {{glossary("Strict mode")}}

> - Dependencies are loaded via static [`import` statements](../en-US/docs/Web/JavaScript/Reference/Statements/import)

Author: ```suggestion
- Dependencies are loaded via static [`import` statements](/en-US/docs/Web/JavaScript/Reference/Statements/import)
```

### Parameters

- `url`
  - : A {{domxref("TrustedScriptURL")}} object or a string representing the URL of the script or module that the worker will execute.
    This must be same-origin with the associated document (`blob:` URLs are same-origin), or a `data:` URL.
    The URL is resolved relative to the current HTML page's location.
- `options` {{optional_inline}}
  - : An object containing option properties that can be set when creating the object instance.
    Available properties are as follows:
    - `credentials`
      - : A string specifying the type of credentials to use for the worker.
        The value can be `omit`, `same-origin`, or _`include`.

>         If not specified, or if type is `classic`, the default used is `same-origin` (only include credentials for same-origin requests)._

Reviewer: I don't understand this. I read it as two separate statements really:

1. "if not specified, the default is same-origin"
2. "if type is `classic`, the default is same-origin" 

But "default" just means "the value used if the option isn't specified", so doesn't (2) just collapse into (1)? Put another way, what's the default if type is `module`?

Does it want to say something like "if type is `classic`, then this option is ignored, and `credentials`are only sent for same-origin requests"?

Author: This predates me - I will need to check.

Author: Addressed. See https://github.com/mdn/content/pull/42073/commits/aa1c134cdad1ed45b97024d9b954c714dbf46b08#r2674765716

  type: "classic",
});

// Construct a module worker
const worker3 = new Worker("worker_module.js", {
  type: "module",
});
```

Module workers and their dependencies are loaded and executed using ECMAScript module semantics:

- Fetched using [CORS](/en-US/docs/Web/HTTP/Guides/CORS)
- All modules are resolved before any code is executed
- Must be served with the media type `Content-Type: text/javascript`
- Executed in {{glossary("Strict mode")}}
- Dependencies are loaded via static [`import` statements](/en-US/docs/Web/JavaScript/Reference/Statements/import)

All same-origin workers can load same-origin scripts by default.
However because module scripts are fetched using CORS, cross-origin scripts must be served with the {{httpheader("Access-Control-Allow-Origin")}} header.
In addition, the if the document has a [Content Security Policy (CSP)](/en-US/docs/Web/HTTP/Guides/CSP), it must allow the origins of imported worker scripts in `worker-src` (with fallback to `script-src` and `default-src` directives).

Classic workers are fetched in `no-cors` mode, executed as scripts, and import other scripts using {{domxref("WorkerGlobalScope.importScripts()")}} (not `import` statements).

> Because they are fetched in `no-cors` mode, cross-origin requests are not blocked by CORS, and the response does not have to include the `Access-Control-Allow-Origin` header.

Reviewer: > @wbamberg So yes, I am saying that classic workers are fetched in no-cors mode. My understanding was that this mean't that any requirements of CORS are ignored - the request will succeed as if CORS headers were not set.

But that's only relevant for cross-origin fetches, right? So if fetching a worker has to be same-origin, it's irrelevant. Put another way, the doc says:

> Because they are fetched in `no-cors` mode, cross-origin requests are not blocked by CORS,

...what are these "cross-origin requests", if fetching a worker has to be same-origin?

> Because they are fetched in `no-cors` mode, cross-origin requests are not blocked by CORS, and the response does not have to include the `Access-Control-Allow-Origin` header.

Author: I have addressed in https://github.com/mdn/content/pull/42073/files#r2674625053

> Unlike for modules, the document [Content Security Policy (CSP)](/en-US/docs/Web/HTTP/Guides/CSP) is not affected by the `worker-src` directive, but it is still affected by sources listed in the `script-src` directive.

Author: My testing shows that if you set a `CSP: worker-src 'self'`, you can load a same-origin script worker that imports a cross origin script. However you can't similarly load the cross-origin module because that is blocked by the CSP.
In other words, scripts ignore worker-src for the imported scripts from script and modules don't.

However a classic worker will respect `script-src`.

Reviewer: It sounds like you're saying that `worker-src` controls the scripts that a worker can load? Not the source of the worker itself? In that case isn't the example here wrong: https://w3c.github.io/webappsec-csp/#directive-worker-src ?

Author: Hoping to get some clarity from Jake in https://github.com/mdn/content/pull/42073/files#r2666249206

Reviewer: Hamish is out for the next two weeks but I promised to try to get this PR merged if I could :).

The question we have is, when a page loads a worker script, does the worker have to be same origin with the loading document? The [page for the `Worker()` constructor](https://developer.mozilla.org/en-US/docs/Web/API/Worker/Worker) says:

> This script must obey the [same-origin policy](https://developer.mozilla.org/en-US/docs/Web/Security/Defenses/Same-origin_policy).

...which seems to me a curious thing to say (I mean, everything has to obey the SOP, it's a browser policy, so what is the substantive force of this statement?). We weren't sure whether this means "the script must be same origin with the loading document". Hamish's tests indicate that it _does_ have to be same origin.

But that raises further questions, such as, what is the `worker-src` CSP directive for, then? The [spec for worker-src](https://w3c.github.io/webappsec-csp/#directive-worker-src) even has an example of how you would use the directive to block workers from a different source, and why would you need that if the worker needs to be same origin?

So we thought we could ask you since you probably know.

> Because they are fetched in `no-cors` mode, cross-origin requests are not blocked by CORS, and the response does not have to include the `Access-Control-Allow-Origin` header.

Author: Attempts to clarify https://github.com/mdn/content/pull/42073/files#r2663618688

Essentially my [reading of the spec](https://html.spec.whatwg.org/multipage/webappapis.html#fetch-a-classic-worker-imported-script) is that all imports on a classic worker are made `no-cors`. It doesn't say that directly, but unlike module requests they are syncronous and don't report errors in the same way as a cors request does. So this part of the spec reads like the `<script>` tag with `no-cors`. Certainly this matches testing.

```suggestion
Classic workers are fetched in `no-cors` mode and are executed as scripts rather than modules.
They can themselves import other scripts using {{domxref("WorkerGlobalScope.importScripts()")}} (not `import` statements), and these are also fetched in `no-cors` mode.
Because imports are fetched in `no-cors` mode,  they can be requested cross-origin even if the server does not set the appropriate CORS headers.
```

  type: "module",
});
```

Module workers and their dependencies are loaded and executed using ECMAScript module semantics:

- Fetched using [CORS](/en-US/docs/Web/HTTP/Guides/CORS)
- All modules are resolved before any code is executed
- Must be served with the media type `Content-Type: text/javascript`
- Executed in {{glossary("Strict mode")}}
- Dependencies are loaded via static [`import` statements](/en-US/docs/Web/JavaScript/Reference/Statements/import)

All same-origin workers can load same-origin scripts by default.
However because module scripts are fetched using CORS, cross-origin scripts must be served with the {{httpheader("Access-Control-Allow-Origin")}} header.
In addition, the if the document has a [Content Security Policy (CSP)](/en-US/docs/Web/HTTP/Guides/CSP), it must allow the origins of imported worker scripts in `worker-src` (with fallback to `script-src` and `default-src` directives).

Classic workers are fetched in `no-cors` mode and are executed as scripts rather than modules.
They can themselves import other scripts using {{domxref("WorkerGlobalScope.importScripts()")}} (not `import` statements), and these are also fetched in `no-cors` mode.

> Because imports are fetched in `no-cors` mode,  they can be requested cross-origin even if the server does not set the appropriate CORS headers.

Author: ```suggestion
Because imports are fetched in `no-cors` mode, scripts can be requested cross-origin even if the server does not set the appropriate CORS headers.
```

### Parameters

- `url`
  - : A {{domxref("TrustedScriptURL")}} object or a string representing the URL of the script the worker will execute.

>     It must obey the same-origin policy.

Author: OK, so this appears to be true. The initial script or module must be same-origin with the document. Note though that the imported scripts from that module need not be.

    - `type`
      - : A string specifying the type of worker to create.
        The value can be `classic` or `module`.
        If not specified, the default used is `classic`.

### Exceptions

- `NetworkError` {{domxref("DOMException")}}
  - : Thrown if the MIME type of the worker script is incorrect. It _should_ always be `text/javascript`
    (for historical reasons [other JavaScript MIME types](/en-US/docs/Web/HTTP/Guides/MIME_types#legacy_javascript_mime_types) may be accepted).
- `SecurityError` {{domxref("DOMException")}}
  - : Thrown if the document is not allowed to start workers, e.g., if the URL has an invalid syntax or if the same-origin policy is violated.
- `SyntaxError` {{domxref("DOMException")}}
  - : Thrown if `url` cannot be parsed.
- `TypeError`
  - : Thrown if the `url` parameter is set with a string when [Trusted Types](/en-US/docs/Web/API/Trusted_Types_API) are [enforced by a CSP](/en-US/docs/Web/API/Trusted_Types_API#using_a_csp_to_enforce_trusted_types) and no default policy is defined.

>     <!-- This is also thrown if the fetched URL cannot be successfully parsed as its indicated type. -->

Author: This was true in other case I copied the error from. Note sure yet if true here.
